File: backend/app/agents/BaseAgent.py
import json
from autogen_core import RoutedAgent, message_handler, MessageContext
from autogen_core.models import SystemMessage, UserMessage
from pydantic import BaseModel
from utils.models import SUPPORTED_MODELS, DEFAULT_MODEL
from utils.log import log, ContentType
from utils.cost import calculate_cost
from store import GLOBAL_PROJECT_ID
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from Models import Agent
from db import get_db_context
from Message import Conversation 

class BaseAgent(RoutedAgent):
    def __init__(self, description: str, system_message: str, response_format: BaseModel) -> None:
        super().__init__(description)
        self._system_message = SystemMessage(content=system_message) 
        self.model = DEFAULT_MODEL

        model_config = SUPPORTED_MODELS[self.model]
        self._model_client = model_config["get_model"](response_format)
        self._model_client_stream = model_config["get_model"](None)
        self.response_format = response_format
        self.url_friendly_id = f"{self.id.key}|{self.id.type}"
        self.description = description

        import asyncio
        asyncio.create_task(self.create_agent())

    async def create_agent(self):
        async with get_db_context() as db:
            # Check if agent already exists for this project
            result = await db.execute(select(Agent).filter(Agent.id == self.url_friendly_id, Agent.project_id == GLOBAL_PROJECT_ID))
            existing_agent = result.scalar_one_or_none()
            
            if existing_agent is None:
                db_agent = Agent(
                    id=self.url_friendly_id,
                    project_id=GLOBAL_PROJECT_ID,
                    model=self.model
                )
                db.add(db_agent)
                await db.commit()
                await db.refresh(db_agent)
                return db_agent
            
            return existing_agent

    async def call(self, prompt: str) -> None:
        log(source=self.id, content=prompt, contentType=ContentType.MESSAGE)
        llm_result = await self._model_client.create(
            messages=[
                self._system_message,
                UserMessage(content=prompt, source=self.id.key),
            ],
        )

        input_tokens = llm_result.usage.prompt_tokens
        output_tokens = llm_result.usage.completion_tokens
        cost = calculate_cost(input_tokens, output_tokens, self.model)

        # Use the context manager to ensure proper connection handling
        async with get_db_context() as db:
            result = await db.execute(select(Agent).filter(Agent.id == self.url_friendly_id, Agent.project_id == GLOBAL_PROJECT_ID))
            db_agent = result.scalar_one()
            db_agent.input_tokens += input_tokens
            db_agent.output_tokens += output_tokens 
            db_agent.cost += cost
            await db.commit()

        log(source=self.id, content=json.dumps({ "cost": cost, "input_tokens": input_tokens, "output_tokens": output_tokens }), contentType=ContentType.INFO)

        content = self.response_format.model_validate_json(llm_result.content)
        return content

    # TODO: Might want to find a way to expose the actual stream to another model/agent.
    async def stream(self, prompt: str) -> None:
        pre_stream_input_tokens = self._model_client.total_usage().prompt_tokens
        pre_stream_output_tokens = self._model_client.total_usage().completion_tokens

        llm_stream = self._model_client_stream.create_stream(
            messages=[
                self._system_message,
                UserMessage(content=prompt, source=self.id.key),
            ],
        )

        output = ""

        async for chunk in llm_stream:
            if (type(chunk) == str):
                log(source=self.id, content=chunk, contentType=ContentType.MESSAGE_STREAM)
                output += chunk

        post_stream_input_tokens = self._model_client.total_usage().prompt_tokens
        post_stream_output_tokens = self._model_client.total_usage().completion_tokens

        input_tokens = post_stream_input_tokens - pre_stream_input_tokens
        output_tokens = post_stream_output_tokens - pre_stream_output_tokens
        cost = calculate_cost(input_tokens, output_tokens, self.model)

        async with get_db_context() as db:
            result = await db.execute(select(Agent).filter(Agent.id == self.url_friendly_id, Agent.project_id == GLOBAL_PROJECT_ID))
            db_agent = result.scalar_one()
            db_agent.input_tokens += input_tokens
            db_agent.output_tokens += output_tokens 
            db_agent.cost += cost
            await db.commit()

        log(source=self.id, content=json.dumps({ "cost": cost, "input_tokens": input_tokens, "output_tokens": output_tokens }), contentType=ContentType.INFO)

        return output

    # All agents support having a conversation with the model directly.
    @message_handler
    async def handle_conversation(self, message: Conversation, ctx: MessageContext) -> None:
        await self.stream(message.chat)

File: backend/src/agents/BaseAgent.ts
import { getModelClient, SUPPORTED_MODELS, DEFAULT_MODEL } from '../utils/models.js';
import { logMessage } from '../utils/logger.js';
import { calculateCost } from '../utils/cost.js';
import { ContentType } from '../constants.js';
import { Agent } from '../models/index.js';
import { GLOBAL_STORE } from '../constants.js';
import { v4 as uuidv4 } from 'uuid';
import { z, ZodObject, ZodRawShape } from 'zod';
import { generateObject, streamText } from 'ai';

// Message structure for memory
interface MemoryMessage {
  role: 'system' | 'user' | 'assistant';
  content: string;
}

/**
 * Base Agent class that all specific agents will extend
 */
export class BaseAgent {
  public readonly agentType: string;
  public readonly description: string;
  public readonly systemMessage: string;
  public model: string;
  public readonly id: string;
  protected memory: MemoryMessage[];

  /**
   * Create a new BaseAgent
   * @param agentType - Type of agent (delegator, builder, etc.)
   * @param description - Human-readable description of the agent
   * @param systemMessage - System message for the LLM
   */
  constructor(agentType: string, description: string, systemMessage: string) {
    this.agentType = agentType;
    this.description = description;
    this.systemMessage = systemMessage;
    this.model = DEFAULT_MODEL;
    this.id = `${this.agentType}|${uuidv4().slice(0, 8)}`;
    this.memory = [];

    // Initialize agent in the database
    void this._initializeAgent();
  }

  /**
   * Initialize agent record in the database
   */
  private async _initializeAgent(): Promise<void> {
    try {
      // Check if agent already exists
      const existingAgent = await Agent.findOne({
        where: {
          id: this.id,
          projectId: GLOBAL_STORE.PROJECT_ID,
        },
      });

      if (!existingAgent) {
        // Create a new agent record
        await Agent.create({
          id: this.id,
          projectId: GLOBAL_STORE.PROJECT_ID,
          model: this.model,
          cost: 0,
          inputTokens: 0,
          outputTokens: 0,
        });
      }
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : 'Unknown error';
      logMessage({
        source: this.id,
        content: `Failed to initialize agent: ${errorMessage}`,
        contentType: ContentType.ERROR,
      });
    }
  }

  /**
   * Update the agent's model
   * @param modelName - New model name
   */
  public async updateModel(modelName: string): Promise<void> {
    try {
      if (!SUPPORTED_MODELS[modelName]) {
        throw new Error(`Model ${modelName} not supported`);
      }

      // Update model in database
      await Agent.update(
        { model: modelName },
        {
          where: {
            id: this.id,
            projectId: GLOBAL_STORE.PROJECT_ID,
          },
        }
      );

      // Update local model reference
      this.model = modelName;

      logMessage({
        source: this.id,
        content: `Model updated to ${modelName}`,
        contentType: ContentType.INFO,
      });
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : 'Unknown error';
      logMessage({
        source: this.id,
        content: `Failed to update model: ${errorMessage}`,
        contentType: ContentType.ERROR,
      });
      throw error;
    }
  }

  /**
   * Call the LLM with a structured output format
   * @param prompt - User prompt
   * @param outputSchema - Zod schema for output validation
   * @returns Structured response from the LLM
   */
  public async call<T extends ZodRawShape>(
    prompt: string,
    outputSchema: ZodObject<T>
  ): Promise<z.infer<ZodObject<T>>> {
    logMessage({
      source: this.id,
      content: prompt,
      contentType: ContentType.MESSAGE,
    });

    try {
      // Add to memory
      this.memory.push({ role: 'user', content: prompt });

      // Prepare messages for the LLM
      const messages: MemoryMessage[] = [
        { role: 'system', content: this.systemMessage },
        ...this.memory.slice(-10), // Keep last 10 messages for context
      ];

      // Get client for current model
      const client = getModelClient(this.model);

      // Call LLM
      const response = await generateObject({
        model: client,
        messages,
        schema: outputSchema,
      });

      // Parse response
      const parsedContent = response.object;
      if (!parsedContent) {
        throw new Error('No content returned from LLM');
      }

      // Add to memory
      this.memory.push({ role: 'assistant', content: JSON.stringify(parsedContent) });

      // Calculate and record usage
      const inputTokens = response.usage?.promptTokens ?? 0;
      const outputTokens = response.usage?.completionTokens ?? 0;
      const cost = calculateCost(inputTokens, outputTokens, this.model);

      // Update agent stats in database
      await this._recordUsage(inputTokens, outputTokens, cost);

      // Log usage information
      logMessage({
        source: this.id,
        content: JSON.stringify({
          cost,
          inputTokens,
          outputTokens,
        }),
        contentType: ContentType.INFO,
      });

      return parsedContent;
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : 'Unknown error';
      logMessage({
        source: this.id,
        content: `Error calling LLM: ${errorMessage}`,
        contentType: ContentType.ERROR,
      });
      throw error;
    }
  }

  /**
   * Stream LLM response
   * @param prompt - User prompt
   * @returns Generator with streamed response
   */
  public async *stream(prompt: string): AsyncGenerator<string, string> {
    logMessage({
      source: this.id,
      content: prompt,
      contentType: ContentType.MESSAGE,
    });

    try {
      // Add to memory
      this.memory.push({ role: 'user', content: prompt });

      // Prepare messages for the LLM
      const messages: MemoryMessage[] = [
        { role: 'system', content: this.systemMessage },
        ...this.memory.slice(-10), // Keep last 10 messages for context
      ];

      // Get client for current model
      const client = getModelClient(this.model);

      // Initialize tracking variables
      let fullResponse = '';
      let usage = { prompt_tokens: 0, completion_tokens: 0 };

      // Stream response
      const { textStream: stream } = await streamText({
        model: client,
        messages,
      });

      for await (const content of stream) {
        if (content) {
          fullResponse += content;

          // Log and yield each chunk
          logMessage({
            source: this.id,
            content,
            contentType: ContentType.MESSAGE_STREAM,
          });

          yield content;

          // Roughly estimate tokens (this is approximate)
          usage.completion_tokens += Math.ceil(content.length / 4);
        }
      }

      // Estimate prompt tokens (this is approximate)
      usage.prompt_tokens = messages.reduce(
        (acc, msg) => acc + Math.ceil(msg.content.length / 4),
        0
      );

      // Add to memory
      this.memory.push({ role: 'assistant', content: fullResponse });

      // Calculate and record usage
      const cost = calculateCost(usage.prompt_tokens, usage.completion_tokens, this.model);

      // Update agent stats in database
      await this._recordUsage(usage.prompt_tokens, usage.completion_tokens, cost);

      // Log usage information
      logMessage({
        source: this.id,
        content: JSON.stringify({
          cost,
          inputTokens: usage.prompt_tokens,
          outputTokens: usage.completion_tokens,
        }),
        contentType: ContentType.INFO,
      });

      return fullResponse;
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : 'Unknown error';
      logMessage({
        source: this.id,
        content: `Error streaming from LLM: ${errorMessage}`,
        contentType: ContentType.ERROR,
      });
      throw error;
    }
  }

  /**
   * Record usage statistics in the database
   * @param inputTokens - Number of input tokens
   * @param outputTokens - Number of output tokens
   * @param cost - Cost of the API call
   */
  private async _recordUsage(
    inputTokens: number,
    outputTokens: number,
    cost: number
  ): Promise<void> {
    try {
      await Agent.increment(
        {
          inputTokens,
          outputTokens,
          cost,
        },
        {
          where: {
            id: this.id,
            projectId: GLOBAL_STORE.PROJECT_ID,
          },
        }
      );
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : 'Unknown error';
      logMessage({
        source: this.id,
        content: `Failed to record usage: ${errorMessage}`,
        contentType: ContentType.ERROR,
      });
    }
  }

  /**
   * Handle a conversation message (to be implemented by subclasses)
   * @param source - Source of the conversation
   * @param message - The conversation message
   */
  public async handleConversation(message: string): Promise<void> {
    // Default implementation just streams the response
    const generator = this.stream(message);
    for await (const _ of generator) {
      // Process each chunk (already logged in stream method)
    }
  }
}
